src/Database/Expression/CommonTableExpression.php

```diff
@@ -0,0 +1,265 @@
+    {
+        $this->name = $name;
+        $this->query = $query;
+    }
+
+    /**
+     * Returns the CTE name.
+     *
+     * @return string
+     */
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    /**
+     * Returns the field names to use for the CTE.
+     *
+     * @return \Cake\Database\ExpressionInterface[]|string[]
+     */
+    public function getFields(): array
+    {
+        return $this->fields;
+    }
+
+    /**
+     * Sets the field names to use for the CTE.
+     *
+     * @param \Cake\Database\ExpressionInterface[]|string[] $fields The field names to use for the CTE.
+     * @return $this
+     */
+    public function setFields(array $fields)
+    {
+        foreach ($fields as $index => $field) {
+            if (is_string($field)) {
+                $fields[$index] = $field = new IdentifierExpression($field);
+            }
+
+            if (!($field instanceof ExpressionInterface)) {
+                throw new InvalidArgumentException(sprintf(
+                    'The `$fields` argument must contain only instances of `%s`, or strings, `%s` given at index `%d`.',
+                    ExpressionInterface::class,
+                    getTypeName($field),
+                    $index
+                ));
+            }
+        }
+
+        $this->fields = $fields;
+
+        return $this;
+    }
+
+    /**
+     * Returns the modifiers to use for the CTE.
+     *
+     * @return \Cake\Database\ExpressionInterface[]|string[]
+     */
+    public function getModifiers(): array
+    {
+        return $this->modifiers;
+    }
+
+    /**
+     * Sets the modifiers to use for the CTE.
+     *
+     * @param \Cake\Database\ExpressionInterface[]|string[] $modifiers The modifiers to use for the CTE.
+     * @return $this
+     */
+    public function setModifiers(array $modifiers)
+    {
+        foreach ($modifiers as $index => $modifier) {
+            if (
+                !($modifier instanceof ExpressionInterface) &&
+                !is_string($modifier)
+            ) {
+                throw new InvalidArgumentException(sprintf(
+                    'The `$modifiers` argument must contain only instances of `%s`, or strings, ' .
+                        '`%s` given at index `%d`.',
+                    ExpressionInterface::class,
+                    getTypeName($modifier),
+                    $index
+                ));
+            }
+        }
+
+        $this->modifiers = $modifiers;
+
+        return $this;
+    }
+
+    /**
+     * Returns the CTE query definition.
+     *
+     * @return \Cake\Database\ExpressionInterface
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * Sets the CTE query definition.
+     *
+     * @param \Cake\Database\ExpressionInterface $query The CTE query definition.
+     * @return $this
+     */
+    public function setQuery(ExpressionInterface $query)
```

Comment:
Should we require a Query instead of an expression? Is there a valid non-query expression for CTE? Should we support `VALUES()` expressions?

Reply:
That might be a little too restrictive, for raw SQL one would usually use `QueryExpression`, and forcing custom expressions to extend `Query` might not be overly appreciated.

`VALUES()` is supported in Postgres and Sqlite, so I'd say it should be supported. There's currently no clean way to to create it with queries or directly via `ValuesExpression` though.

Reply:
We could support `QueryExpression|ValuesExpression` in the docblock, but this is good for now.